qiskit/synthesis/linear/__init__.py

```diff
@@ -12,7 +12,7 @@
 
 """Module containing cnot circuits"""
 
-from .graysynth import graysynth, synth_cnot_count_full_pmh
+from .cnot_synth import synth_cnot_count_full_pmh
```

Comment:
This makes the public function `graysynth` no longer accessible from `qiskit/synthesis/linear` without a warning.  We need to keep that accessible during any period of deprecation.  If `synth_cnot_phase_aam` is identical to `graysynth`,that could be done by re-importing it from `synthesis.linear_phase` to here with a new name.

If it's not identical, then we need to keep the function in existence.  Either way, the function isn't eligible for deprecation until the alternative has existed for at least one minor release of Terra, so in this case, `qiskit.synthesis.linear.graysynth` needs to remain accessible with no warning for Terra 0.24, and can gain a `DeprecationWarning` in Terra 0.25 if the alternative is released in 0.24.

Reply:
this code was temporarily moved there and did not appear in the docs. I'm not sure if we should deprecate it.

Reply:
If it was here in a release version of Terra, it needs formal deprecation.  Much as we'd like our complete API surface to be exactly what's in the docs, in practice people read our source code and rely on things being where they were originally put, so we do proper deprecations for everything.

It looks to me that `graysynth` was here in Terra 0.23, so it needs a deprecation before we remove it.  If there are no cyclic dependencies, you can satisfy it by just doing
```python
from qiskit.synthesis.linear_phase import synth_cnot_count_aam as graysynth
```
or if there are cyclic dependencies, you can do
```python
def __getattr__(name):
    if name == "graysynth":
        from qiskit.synthesis.linear_phase import synth_cnot_count_aam

        return synth_cnot_count_aam
    raise AttributeError(f"module {__name__} has no attribute {name}")
```

Reply:
I tried your suggestion, and still go pylint import errors.
So I just imported the file and added some pylint disables. Hope it's OK now.

Reply:
Yeah, that's fine thanks.  Pylint errors were to be expected, since it is a true cyclic dependency - my concern was about the module failing to import at all.  As long as that works (which it does), we're ok.